### PR TITLE
Ab#81881 request comments

### PR DESCRIPTION
**Pull request by TC-Cory**

### CODE REVIEW

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        [Inject]
+        public IHttpContextAccessor httpContextAccessor { get; set; }
+        [Inject]
+        IStringLocalizer<Shared.Common> Localizer { get; set; }
+        private SortOption SortOption { get; set; } = new SortOption();
+        public EditContext EditContext { get; set; }
+        private List<RequestCommentInfo> workComments { get; set; }
+        public RequestComment RequestComment { get; set; } = new RequestComment();
+        private int maxShownComments { get; set; } = 5;
+        private bool areAllCommentsShown { get; set; } = false;
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:46]:
```suggestion
        [Parameter]
        public bool ReadOnly { get; set; }
        
        [Parameter]
        public EventCallback<string> SubmitComment { get; set; }
        
        [Parameter]
        public EventCallback<List<RequestCommentInfo>> WorkCommentsChanged { get; set; }
        
        [Inject]
        public IHttpContextAccessor httpContextAccessor { get; set; }
        
        [Inject]
        public IStringLocalizer<Shared.Common> Localizer { get; set; }
        
        public EditContext EditContext { get; set; }
        
        public RequestComment RequestComment { get; set; } = new RequestComment();
        
        private SortOption SortOption { get; set; } = new SortOption();
        
        private List<RequestCommentInfo> workComments { get; set; }
        
        private int maxShownComments { get; set; } = 5;
        
        private bool areAllCommentsShown { get; set; } = false;
```

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            this.setMaxToCurrentCount();
+            base.OnAfterRender(firstRender);
+        }
+
+        public async void SaveComment()
+        {
+            if (this.EditContext.Validate())
+            {
+                await SubmitComment.InvokeAsync(this.RequestComment.Comment);
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:94]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        public async void SaveComment()
+        {
+            if (this.EditContext.Validate())
+            {
+                await SubmitComment.InvokeAsync(this.RequestComment.Comment);
+               
+            }
+                this.maxShownComments += 1;
+            if(this.maxShownComments > this.WorkComments.Count)
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:97]:
```suggestion
            }
            this.maxShownComments += 1;
            if(this.maxShownComments > this.WorkComments.Count)
```

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            {
+                this.maxShownComments = this.WorkComments.Count;
+            }
+            
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:102]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            {
+                this.WorkComments = WorkComments.OrderByDescending(o => o.CreatedDateUTC).ToList();
+            }
+            else
+            {
+                this.WorkComments = WorkComments.OrderBy(o => o.CreatedDateUTC).ToList();
+            }
+            StateHasChanged();
+        }
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:120]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+        private void LoadMoreComments()
+        {
+            this.maxShownComments += 10;
+            this.setMaxToCurrentCount();
+            StateHasChanged();
+        }
+        private void setMaxToCurrentCount()
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:128]:
```suggestion
        }
        
        private void setMaxToCurrentCount()
```

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            }
+        }
+        private bool IsCommentListEmpty() => !(this.WorkComments == null || !this.WorkComments.Any());
+    }
+
+    public partial class SortOption
+    {
+        public string Id { get; set; }
+        public string Text { get; set; }
+    }
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:143]: Is there a reason this is a partial class? We should move this to its own file.

File: src/CSF.SRDashboard.Client/PageValidators/CommentValidator.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace CSF.SRDashboard.Client.PageValidators
+{
+    public class CommentValidator : AbstractValidator<RequestComment>
+    {
+        public CommentValidator() {
+            this.RuleFor(i => i.Comment).NotEmpty().WithMessage("Please enter a comment before submitting");
+                 }
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/PageValidators/CommentValidator.cs:14]:
```suggestion
        public CommentValidator() {
            this.RuleFor(i => i.Comment).NotEmpty().WithMessage("Please enter a comment before submitting");
        }
```

File: src/CSF.SRDashboard.Client/Pages/CreateNewRequest.razor.cs [hunk 9]
```diff
@@ -59,7 +59,7 @@ public partial class CreateNewRequest
 
         public RequestValidator validator = new RequestValidator();
 
-
+        public List<RequestCommentInfo> WorkComments { get; set; } = new List<RequestCommentInfo>();
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/CreateNewRequest.razor.cs:62]:
```suggestion
        public List<RequestCommentInfo> WorkComments { get; set; } = new List<RequestCommentInfo>();
        
```

File: src/CSF.SRDashboard.Client/Pages/CreateNewRequest.razor.cs [hunk 10]
```diff
@@ -180,6 +181,38 @@ public string FindPreferredLanguage(MpdisApplicantDto Applicant)
...
+                RequestCommentInfo tempWorkComment = new RequestCommentInfo()
+                {
+                    Comment = commentText,
+                    CreatedDateUTC = DateTime.UtcNow,
+                    CreatedBy = this.State.LoggedInUser
+                };
+                this.WorkComments.Insert(0, tempWorkComment);
+            }
+           
+        }
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/CreateNewRequest.razor.cs:215]:
```suggestion
        }
        
```

File: src/CSF.SRDashboard.Client/Pages/EditNewRequest.razor.cs [hunk 11]
```diff
@@ -202,6 +205,47 @@ public async void SaveChanges()
...
+                this.WorkComments.Insert(0, new RequestCommentInfo()
+                {
+                    Comment = commentText,
+                    CreatedBy = this.State.LoggedInUser,
+                    CreatedDateUTC = DateTime.UtcNow
+                });
+                await this.WorkLoadService.AddWorkItemComment(workCommentToInsert);
+
+            
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/EditNewRequest.razor.cs:229]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs [hunk 12]
```diff
@@ -42,12 +45,19 @@ public partial class ViewRequestDetails
         [Inject]
         public NavigationManager NavigationManager { get; set; }
 
+        [Inject]
+        public IHttpContextAccessor httpContextAccessor { get; set; }
+
+        [Inject]
+        public SessionState State { get; set; }
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs:52]:
```suggestion
        public SessionState State { get; set; }
        
```

File: src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs [hunk 13]
```diff
@@ -108,7 +120,65 @@ protected async override Task OnInitializedAsync()
 
             StateHasChanged();
         }
+        public async void InsertCommentOnRequest(string commentText)
+        {
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs:125]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs [hunk 14]
```diff
@@ -108,7 +120,65 @@ protected async override Task OnInitializedAsync()
...
+            if (!string.IsNullOrEmpty(commentText))
+            {
+                this.WorkComments.Insert(0, new RequestCommentInfo()
+                {
+                    Comment = commentText,
+                    CreatedBy = this.State.LoggedInUser,
+                    CreatedDateUTC = now,
+                    CreatedDateFormatted = Constants.DateFormat(now, language, null, Localizer)
+                });
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs:146]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs [hunk 15]
```diff
@@ -108,7 +120,65 @@ protected async override Task OnInitializedAsync()
...
+                {
+                    Comment = commentText,
+                    CreatedDateUTC = DateTime.UtcNow,
+                    WorkItemId = this.RequestId,
+                    CreatedBy = this.State.LoggedInUser
+                };
+
+                await this.WorkLoadService.AddWorkItemComment(workCommentToInsert);
+
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs:158]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs [hunk 16]
```diff
@@ -108,7 +120,65 @@ protected async override Task OnInitializedAsync()
...
+            }
+
+            StateHasChanged();
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs:162]:
```suggestion
```

File: src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs [hunk 17]
```diff
@@ -108,7 +120,65 @@ protected async override Task OnInitializedAsync()
...
+        }
+
+
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Pages/ViewRequestDetails.razor.cs:165]:
```suggestion
```

**Review by AlexandreSanscartier — CHANGES_REQUESTED** [review]
Just some styling changes, other than that looks good!

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        private void setMaxToCurrentCount()
+        {
+            if (this.maxShownComments >= this.WorkComments.Count)
+            {
+                this.maxShownComments = this.WorkComments.Count;
+                this.areAllCommentsShown = true;
+            }
+        }
+
+        private bool isCommentListEmpty() => !(this.WorkComments == null || !this.WorkComments.Any());
```
Comment (AlexandreSanscartier) [inline src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs:146]: We need to change this statement to the following or fix the variable name. 
```suggestion
        private bool isCommentListEmpty => this.WorkComments== null || !this.WorkComments.Any();
```

**Review by AlexandreSanscartier — CHANGES_REQUESTED** [review]

File: src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs (hunk 18 above)
Comment (TC-Cory) [reply src/CSF.SRDashboard.Client/Components/Tables/WorkloadRequest/RequestCommentComponent.razor.cs]: Done

**Pull request merged**